src/api/script-cache.ts

```diff
@@ -0,0 +1,106 @@
+import { Link } from "expression/link";
+import { Datastore } from "index/datastore";
+import { Result } from "./result";
+import { MarkdownCodeblock, MarkdownSection } from "index/types/markdown";
+import { DatacoreJSRenderer, ScriptLanguage, asyncEvalInContext } from "ui/javascript";
+import { DatacoreLocalApi } from "./local-api";
+import { Fragment, h } from "preact";
+export interface DatacoreScript {
+    language: ScriptLanguage;
+    code: string | null;
+    id: string;
+    state: LoadingState;
+    source: string;
+}
+export const enum LoadingState {
+    UNDEFINED = -1,
+    LOADING,
+    LOADED,
+}
+/** A simple caching script loader that can load any DAG of script dependencies. */
+export class ScriptCache {
+    /** A static placeholder placed into the `scripts` section to catch recursive loading loops. */
+    private static readonly LOADING_SENTINEL: unique symbol = Symbol("CURRENTLY_LOADING");
+    private static SCRIPT_TAGS: string[] = ["javascript", "typescript", "ts", "js", "tsx", "jsx"];
+
+    public constructor(private store: Datastore) {}

```

Comment:
I would move the loaded scripts state here. I would also make it a map instead of a plain list - we want unique keys here (by the path/link of each script we are loading).